### PR TITLE
[FLINK-33441] Implement restore tests for ExecUnion node

### PR DESCRIPTION
**Pull request by bvarghese1**
## What is the purpose of the change

Add restore tests for ExecUnion node

## Verifying this change

This change added tests and can be verified as follows:

  - *Added restore tests for ExecUnion node which verifies the generated compiled plan with the saved compiled plan*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java [hunk 1]
```diff
@@ -91,7 +91,8 @@ public EnumSet<TestKind> supportedSetupSteps() {
         return EnumSet.of(
                 TestKind.FUNCTION,
                 TestKind.SOURCE_WITH_RESTORE_DATA,
-                TestKind.SINK_WITH_RESTORE_DATA);
+                TestKind.SINK_WITH_RESTORE_DATA,
+                TestKind.SINK_WITH_DATA);
```
[the remark this answers is not in the document]
Comment (bvarghese1) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java]: Yeah fixed this.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/UnionTestPrograms.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                                                    6,
+                                                    "hello there",
+                                                    LocalDateTime.of(
+                                                            2023, 12, 19, 01, 01, 01, 123)))
+                                    .build())
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("source_t2")
+                                    .addSchema("d BIGINT", "e INT NOT NULL")
+                                    .producedBeforeRestore(Row.of(420L, 1), Row.of(421L, 2))
+                                    .producedAfterRestore(Row.of(500L, 3), Row.of(421L, 2))
```
Comment (jnh5y) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/UnionTestPrograms.java:106]: Nice demonstration of UNION ALL.

**Review by jnh5y — APPROVED** [review]
LGTM.  In particular:

1. All existing tests are ported.
2. An additional test is added.
3. Test names start with "union-" / I don't see any other common nits.

**Review by dawidwys — APPROVED** [review]
LGTM

**Pull request merged**